src/monitormanager.h

```diff
@@ -30,6 +30,8 @@ class MonitorManager : public ChildByIndex<HSMonitor> {
     std::function<int(Input, Output)> byFirstArg(HSMonitorCommand cmd);
     // relayout the monitor showing this tag, if there is any
     void relayoutTag(HSTag* tag);
+    int removeMonitor(Input input, Output output);
+    void removeMonitor(HSMonitor* monitor);
```

Comment:
My incoherent ramblings on this:

Yes, there are many method call chains like this in the code, some of them even go through several hops (Thorsten is aware of this problem). It is something that was always a thorn in my flesh (Dorn im Auge) for me, but on the other hand: the (Input, Output) interface travels deep down into the code often as a necessity to allow feedback to the user from there. Now it could still be worthwhile to provide a generic solution that avoids overloads where this is not the case.

Note that what is not visible in the code yet is that many of the commands can be aliased into attribute changes and object actions (right now missing in the code!). An action can be something like `clients.focus.close`. An attribute change like setting `foo.name` can replace a rename command.

I would argue that removeMonitor is an example where each monitor object could have a remove action and then the command would alias to that. The remove action could be written as a lambda. Checks like valid name or index don't need to be in the code anymore because the object tree will check this. `string_to_monitor_index` would be used by the function that establishes the alias. Still a bit cumbersome.

Reply:
To clarify my ramblings, we could end up in a situation where commands are either aliases to attributes/actions or are done outside a specific object anyways (see rootcommands.cpp). I woud would like it as a somewhat pure solution, if it really works out as expected.
This goal is still a personal one, while we did already agree on aliasing commands to attributes/actions in cases where it _obviously_ makes sense.

Reply:
This all sounds very reasonable.

I share your conclusion that some commands need to "go deep" for useful feedback, so that it's probably not workable to try to fully separate the parsing from the doing in all commands.

On the other hand, the user feedback could be improved in terms of, for instance, providing a proper help string (usage) on failure. So, unrelated to most of what you said, I'm hoping that we could at some point provide the command functions with an interface similar to Python's argparse. Using that, the commands could then generically declare how they want to be invoked (number of args, metavars, which are optional, etc.), but still work with the Input object further down the road to provide useful error messages.

I don't intend to reimplement argparse in C++, but if there's something good available, maybe we could use that to our advantage.

Reply:
Placing actions isn't easy as you notice (`monitors.remove 1` vs `monitors.1.remove`; the former is more reasonable w.r.t. implementation, the latter is more intuitive for the user, because it is clearly an action *on* the monitor 1).

We don't even need something complex like argparse, because the commands only have positional arguments and not named arguments (even though the positional arguments sometimes start with a `--`, e.g. for `keyunbind`).

In the hlwm code, the command member functions often have the suffix `Cmd` and so the overloading is avoided. Also note that `BIND_OBJECT` and `byFirstArg` are not two patterns but simply to different things, since the latter is specific to some objects and already parses the first argument)